### PR TITLE
Verisk coding exercise

### PR DESCRIPTION
**Pull request by daweili1215**

### CODE REVIEW

File: Verisk/linked_list.hpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  void clean_up() {
+    while (head) {
+      list_node<T>* tmp = head;
+      head = head->next;
+      delete tmp;
+    }
+  }
+
+  void insert(T value) {
+    auto node = new list_node<T>(value);
```
Comment (TimothyBanks) [inline Verisk/linked_list.hpp:52]: Use uniform initialization.

File: Verisk/multithreads.cpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+void print_content(uint32_t* counts) {
+  for (int i = 0; i < RANGE; ++i) {
+    if (counts[i] > 0) 
+      cout << "counts of " << i << ": " << counts[i] << endl;
+  }
+}
+
+int main() {
+  uchar data[] = {0,1,2,3,4,5,4,3,2,1,100,100,255,255};
```
Comment (TimothyBanks) [inline Verisk/multithreads.cpp:60]: Use a standard container such as `std::vector` or `std::array`

File: Verisk/multithreads.cpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  int data_size = sizeof(data) / sizeof(uchar);
+  uint32_t histogram[256] = {0};
```
Comment (TimothyBanks) [inline Verisk/multithreads.cpp:62]: Use a standard container such as `std::vector` or `std::array`

Comment (TimothyBanks) [reply Verisk/multithreads.cpp:62]: I see, you are forced to use a specific function signature.

File: Verisk/multithreads.cpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
+#include <iostream>
+#include <vector>
+#include <thread>
+#include <mutex>
+
+using namespace std;
+using uchar = unsigned char;
+std::mutex mtx;
```
Comment (TimothyBanks) [inline Verisk/multithreads.cpp:8]: Declare this closer to where it is being used.

File: Verisk/multithreads.cpp [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+Given an array of uchars, implement the multithreading application to determine the frequency of distinct
+values in the array. From the function prototype below, accessing result[0] would give the frequency of the
+value 0, result[1] would give the frequency of value 1, etc.  
+dataCount: Length of inputData
+result: Output result
+CreateHistogram (uchar* inputData, int dataCount, uint* result){}
+*/
+
+
+const int RANGE = 256;
```
Comment (TimothyBanks) [inline Verisk/multithreads.cpp:21]: `constexpr`

File: Verisk/multithreads.cpp [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+const int THREAD_LIMIT = 4;
```
Comment (TimothyBanks) [inline Verisk/multithreads.cpp:22]: `constexpr`

File: Verisk/multithreads.cpp (hunk 2 above)
Comment (TimothyBanks) [reply Verisk/multithreads.cpp:60]: I see, you are forced to use a specific signature.

File: Verisk/multithreads.cpp [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  std::vector<std::thread> threads;
+  auto chunk_length = dataCount / THREAD_LIMIT;
+  
+  auto thread_worker = [&](auto start, auto end) {
+    uint32_t counts[RANGE] = {0};
+    
+    for (auto i = start; i < end; ++i)
+      ++counts[inputData[i]];
+    
+    std::lock_guard<std::mutex> lock(mtx);
```
Comment (TimothyBanks) [inline Verisk/multithreads.cpp:37]: The lock seems a bit heavy handed.  One thing to think about here is that you already are instantiating memory to hold the local results.  You could do something like this which is a common parallel pattern:

```
uint32_t counts[THREAD_LIMIT][RANGE];

auto thread_worker = [&](auto i) {
  counts[i] = {0};

  auto start = i * chunk_length;
  auto end = start + chunk_length;
  end = end >= dataCount ? dataCount : end;

  for (auto i = start; i < end; ++i)
    ++counts[i][inputData[i]];
}

for (int i = 0; i < THREAD_LIMIT; ++i) {
  threads.emplace_back(thread_worker, i;
}

for (auto& t : threads)
  t.join();

for (auto i = 0; i < RANGE; ++i) {
  for (auto j = 0; j < THREAD_LIMIT; ++j) {
    results[i] += counts[j][i];
  }
}
```

Comment (TimothyBanks) [reply Verisk/multithreads.cpp:37]: With this approach you would spend less time iterating over the results array, forgo the need for a mutex and the memory requirements would be the same.

File: Verisk/linked_list.hpp [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+Write a template class that implements a single-linked list.
+Write two methods of this class, both of which will reverse the order of the elements in the list. One
+function should use a recursive approach; the second function should use an iterative approach.
+For both functions, do not create a new list items; must instead re-use existing list items.
+*/
+
+template <typename T>
+struct list_node {
+  T val;
+  list_node* next;
```
[the remark this answers is not in the document]
Comment (daweili1215) [reply Verisk/linked_list.hpp:18]: `  
  void recursive_reverse() {
    head = reverse_helper(std::move(head));
  }
node_ptr reverse_helper(node_ptr node) {
    if (!node) {
      return nullptr;
    }
      
    if (!node->next) {
      return node;
    }
    
    node_ptr new_head = reverse_helper(std::move(node->next));
    node->next->next = std::move(node);
    node->next = nullptr;
    return new_head;
  }
  `
  How come this creates a segmentation fault?

[the remark this answers is not in the document]
Comment (daweili1215) [reply Verisk/linked_list.hpp]: Never mind. Just resolved.

File: Verisk/multithreads.cpp (hunk 7 above)
Comment (daweili1215) [reply Verisk/multithreads.cpp]: Thanks for your suggestions on this interview.

Comment (TimothyBanks) [reply Verisk/multithreads.cpp]: This is the fork and join pattern. I had trouble remembering the name last night.

**Pull request merged**